src/backend/utils/misc/queryenvironment.c

```diff
@@ -978,8 +1111,469 @@ void ENRDropEntry(Oid id)
+				{
+					/*
+					* We skip invalidating IndexRelationId on drop (ie committed table, index created in transaction) 
+					* because it will be taken care of when the index itself is dropped - otherwise we risk
+					* throwing an error because the entry is already wiped away.
+					*/
+					Form_pg_index idx_form = (Form_pg_index) GETSTRUCT(uncommitted_tup->tup);
+					EphemeralNamedRelation tmp_enr = get_ENR_withoid(queryEnv, idx_form->indrelid, ENR_TSQL_TEMP);
+					if (tmp_enr && tmp_enr->md.is_committed)
+						skip_cache_inval = true;
+				}
+
+				rel = relation_open(uncommitted_tup->catalog_oid, AccessExclusiveLock);
+
+				switch(uncommitted_tup->optype)
+				{
+					/* Peform the opposite operation as the saved optype. */
+					case ENR_OP_ADD:
+						_ENR_tuple_operation(rel, uncommitted_tup->tup, ENR_OP_DROP, skip_cache_inval);
+						break;
+					case ENR_OP_UPDATE:
+						ENRupdateTuple(rel, uncommitted_tup->tup);
+						break;
+					case ENR_OP_DROP:
+						_ENR_tuple_operation(rel, uncommitted_tup->tup, ENR_OP_ADD, skip_cache_inval);
+						break;
+				}
+				relation_close(rel, AccessExclusiveLock);
+			}
+		}
+
+		in_enr_rollback = false;
+
+		ENRDeleteUncommittedTupleData(InvalidSubTransactionId, enr);
+	}
+
+	/* Finally, clean up any ENRs that must be deleted. */
+	foreach(lc, enrs_to_drop)
+	{
+		EphemeralNamedRelation enr = (EphemeralNamedRelation) lfirst(lc);
+		queryEnv->namedRelList = list_delete(queryEnv->namedRelList, enr);
+		pfree(enr->md.name);
+		pfree(enr);
+	}
+	MemoryContextSwitchTo(oldcxt);
+}
+
+/* This is used when rolling back a subtransaction. */
+/* It is similar to ENRRollback */
+void ENRRollbackSubtransaction(SubTransactionId subid, QueryEnvironment *queryEnv)
+{
+	ListCell 	*lc;
+	List *enrs_to_drop = NIL;
+	MemoryContext oldcxt;
+
+	if (!queryEnv)
+		return;
+
+	oldcxt = MemoryContextSwitchTo(queryEnv->memctx);
+
+	/* Add back any ENRs that must be restored. */
+	foreach (lc, queryEnv->dropped_namedRelList)
+	{
+		EphemeralNamedRelation enr = (EphemeralNamedRelation) lfirst(lc);
+
+		Assert(enr->md.enrtype == ENR_TSQL_TEMP);
+		Assert(!enr->md.is_table_variable);
+
+		/* If this was dropped in the current subtransaction, then we must restore it. */
+		if (enr->md.dropped_subid != InvalidSubTransactionId && enr->md.dropped_subid >= subid) // 4
+		{
+			queryEnv->namedRelList = lappend(queryEnv->namedRelList, enr);		
+			queryEnv->dropped_namedRelList = foreach_delete_current(queryEnv->dropped_namedRelList, lc);
+			enr->md.dropped_subid = InvalidSubTransactionId;
+		}
+	}
+
+	/*
+	 * Loop through the registered ENRs to process each.
+	 */
+	foreach(lc, queryEnv->namedRelList)
+	{
+		Relation rel;
+		EphemeralNamedRelation enr = (EphemeralNamedRelation) lfirst(lc);
+
+		if (enr->md.enrtype != ENR_TSQL_TEMP || enr->md.is_table_variable)
+			continue;
+
+		if (!enr->md.is_committed && enr->md.created_subid >= subid)
```

Comment:
same as above, how will single sub id let us know which ENRs to drop?

Reply:
See comment above